apps/blockchain/lib/blockchain/ethash.ex

```diff
@@ -0,0 +1,56 @@
+defmodule Blockchain.Ethash do
+  @moduledoc """
+  This module contains the logic found in Appendix J of the
+  yellow paper concerning the Ethash implementation for POW.
+  """
+
+  @j_epoch 30_000
+  @j_datasetinit :math.pow(2, 30) |> round
+  @j_datasetgrowth :math.pow(2, 23) |> round
+  @j_mixbytes 128
+  @j_cacheinit :math.pow(2, 24) |> round
+  @j_cachegrowth :math.pow(2, 17) |> round
+  @j_hashbytes 64
+
+  def epoch(block_number) do
+    div(block_number, @j_epoch)
+  end
+
+  def dataset_size(epoch, cache \\ Ethash.Cache.data_sizes()) do
+    case Enum.at(cache, epoch) do
+      nil -> calculate_dataset_size(epoch)
+      found_data_size -> found_data_size
+    end
+  end
+
+  defp calculate_dataset_size(epoch) do
+    prime(@j_datasetinit + @j_datasetgrowth * epoch - @j_mixbytes, @j_mixbytes)
+  end
+
+  def cache_size(epoch, cache \\ Ethash.Cache.cache_sizes()) do
+    case Enum.at(cache, epoch) do
```

Comment:
would this work?

    Enum.at(cache, epoch, calculate_cache_size(epoch))

Reply:
I believe that wouldn't short-circuit evaluation

Reply:
We were actually wondering about that. I'll take another look to see if we could do that if it short-circuits evaluation. I think it would be nicer

Reply:
`Enum.at(cache, epoch) || calculate_cache_size(epoch)` may make the short circuiting behavior more obvious.

Reply:
☝️ we added that in f827391ca05a1bd35352b66092040fc8aea17071. @masonforest your suggestion unfortunately executes `calculate_cache_size(epoch)` every time, so it doesn't short-circuit it (like Geoff mentioned)